pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/ServerRequestPlanVisitor.java

```diff
@@ -126,8 +126,7 @@ public static ServerPlanRequestContext build(MailboxService<TransferableBlock> m
     QUERY_OPTIMIZER.optimize(pinotQuery, tableConfig, schema);
 
     // 2. set pinot query options according to requestMetadataMap
-    pinotQuery.setQueryOptions(
-        ImmutableMap.of(CommonConstants.Broker.Request.QueryOptionKey.TIMEOUT_MS, String.valueOf(timeoutMs)));
+    updateQueryOptions(pinotQuery, requestMetadataMap, timeoutMs, traceEnabled);
```

Comment:
do we want to support stage based query options as well?

Reply:
stage-based query options are also added. they are merged into requestMetadataMap during dispatch.